### PR TITLE
dr/tp

### PR DESCRIPTION
**Pull request by sebwolf-de**
Implement thermal pressurization in c++.
Failing Pipeline is connected to #520. Draft PR until #520 is fixed on the master or #513 is merged into dr/cpp.
Anyways, early review is appreciated ;-)

### CODE REVIEW

File: gitlab-cpu-pipeline.yml [hunk 1]
```diff
@@ -42,13 +42,15 @@ ci_builder_pre:
...
         - pip3 install sphinx_rtd_theme
+        - pip3 install 'seissolxdmf>=0.1.0'
+        - pip3 install h5py
         - tar -xzf netcdf-4.6.1.tar.gz
         - cd netcdf-4.6.1 && CC=h5pcc ./configure --prefix=/usr --enable-shared=no --disable-dap && make && sudo make install && cd ..
         - unzip libxsmm-1.17.zip
         - cd libxsmm-1.17 && make generator && sudo cp bin/libxsmm_gemm_generator /usr/bin && cd ..
         - git clone https://github.com/uphoffc/ImpalaJIT.git &&
           mkdir impala_build && cd impala_build && cmake ../ImpalaJIT && sudo make -j $(nproc) install && cd ..
-        - git clone --depth 1 --branch v1.0.0 https://github.com/SeisSol/easi.git &&
```
Comment (ravil-mobile) [inline gitlab-cpu-pipeline.yml:51]: Maybe it is better to specify a specific version? What do you think or there is a reason to not to do it?

File: src/DynamicRupture/FrictionLaws/RateAndState.h [hunk 2]
```diff
@@ -22,6 +22,8 @@ class RateAndStateBase : public BaseFrictionLaw<RateAndStateBase<Derived>> {
   real (*sl0)[misc::numPaddedPoints];
   real (*stateVariable)[misc::numPaddedPoints];
 
+  TPMethod tpMethod;
+
   // TU 7.07.16: if the SR is too close to zero, we will have problems (NaN)
   // as a consequence, the SR is affected the AlmostZero value when too small
   const real almostZero = 1e-45;
```
Comment (ravil-mobile) [inline src/DynamicRupture/FrictionLaws/RateAndState.h:29]: `const real almostZero = 1e-45;`
maybe we can use `constexpr`

File: src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+ * Logarithmic gridpoints as defined in Noda&Lapusta (14). These are the \f$\hat{l}\f$ for
+ * ThermalPressurization.
+ */
+template <size_t N>
+class GridPoints {
+  public:
+  GridPoints() {
+    for (size_t i = 0; i < N; ++i) {
+      values[i] =
+          misc::tpMaxWavenumber * std::exp(-misc::tpLogDz * (misc::numberOfTPGridPoints - i - 1));
```
Comment (ravil-mobile) [inline src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h:23]: maybe `tpMaxWaveNumber`

File: src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  real (*theta)[misc::numPaddedPoints][misc::numberOfTPGridPoints];
+  real (*sigma)[misc::numPaddedPoints][misc::numberOfTPGridPoints];
+  real (*thetaTmpBuffer)[misc::numPaddedPoints][misc::numberOfTPGridPoints];
+  real (*sigmaTmpBuffer)[misc::numPaddedPoints][misc::numberOfTPGridPoints];
+  real (*halfWidthShearZone)[misc::numPaddedPoints];
+  real (*hydraulicDiffusivity)[misc::numPaddedPoints];
+  real (*faultStrength)[misc::numPaddedPoints];
+
+  public:
+  /**
```
Comment (ravil-mobile) [inline src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h:120]: Let's list all public methods on the top, then protected and then private. I personally don't like the following style:
```
class A {
public:
  void foo1();
private:
  int a{};
public:
  void bar1();
protected:
  float b{};
};
```

I usually prefer this one:
```
class A {
public:
  void foo1();
  void bar1();
protected:
  float b{};
private:
  int a{};
};
```

My rationale is the following: we start reading from the top to bottom, and we usually want to know what an object of a class can do (i.e., what is the interface). Then if we want to extend a class we can scroll further and examine what we can/need to override. If we really need to change the class then, yes, we will scroll till the bottom to examine the internals of the class

File: src/DynamicRupture/Misc.h [hunk 5]
```diff
@@ -25,6 +25,13 @@ constexpr size_t leadDim() noexcept {
 static constexpr inline size_t numPaddedPoints = leadDim<init::QInterpolated>();
 static constexpr inline size_t numQuantities = misc::dimSize<init::QInterpolated, 1>();
 
+/**
+ * Constants for Thermal Pressurization
+ */
+static constexpr size_t numberOfTPGridPoints = 60;
+static constexpr real tpLogDz = 0.3;
+static constexpr real tpMaxWavenumber = 10.0;
```
Comment (ravil-mobile) [inline src/DynamicRupture/Misc.h:33]: This is good! I always wanted to move `numberOfTPGridPoints` to `Misc.h`

File: src/Solver/Interoperability.cpp [hunk 6]
```diff
@@ -408,7 +408,9 @@ double c_interoperability_M2invDiagonal(int no) {
 
   extern void f_interoperability_setFrictionOutputStrength(void*  i_domain, int i_face, real* strength);
 
-  extern void f_interoperability_setFrictionOutputInitialStress(void*  i_domain,
+extern void f_interoperability_setFrictionOutputThermalPressurization(void*  i_domain, int i_face, real* fluidPressure, real* fluidTemperature);
+
+extern void f_interoperability_setFrictionOutputInitialStress(void*  i_domain,
                                                                 int i_face,
                                                                 real* i_initialStressInFaultCS,
                                                                 real* i_bulkXX,
```
Comment (ravil-mobile) [inline src/Solver/Interoperability.cpp:416]: Please, check the indentation. Make sure that it matches with the rest of the file. I would suggest to write all arguments in a new line as follows:
```
extern void f_interoperability_setFrictionOutputThermalPressurization(void*  i_domain,
                                                                      int i_face,
                                                                      real* fluidPressure,
                                                                      real* fluidTemperature);

```

File: src/Solver/Interoperability.h [hunk 7]
```diff
@@ -477,6 +467,11 @@ class seissol::Interoperability {
                                            real  (*strength)[init::QInterpolated::Stop[0]]
   );
 
+  void copyFrictionOutputToFortranThermalPressurization(unsigned ltsFace, unsigned meshFace,
+                                           real  (*fluidPressure)[dr::misc::numPaddedPoints],
+                                           real  (*fluidTemperature)[dr::misc::numPaddedPoints]
+  );
```
Comment (ravil-mobile) [inline src/Solver/Interoperability.h:473]: please, align the argument list

File: gitlab-cpu-pipeline.yml (hunk 1 above)
Comment (sebwolf-de) [reply gitlab-cpu-pipeline.yml:51]: I just wanted to use the latest version of easi. We can specify a specific version here, but then we might need to adapt the pipeline script, when there is an easi update.

File: src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+template <size_t N>
+class InverseFourierCoefficients {
+  public:
+  constexpr InverseFourierCoefficients() {
+    GridPoints<N> localGridPoints;
+
+    for (size_t i = 1; i < N - 1; ++i) {
+      values[i] = std::sqrt(2 / M_PI) * localGridPoints[i] * misc::tpLogDz;
+    }
+    values[0] = std::sqrt(2 / M_PI) * localGridPoints[0] * (1 + misc::tpLogDz);
```
Comment (Thomas-Ulrich) [inline src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h:44]: The missing 0.5  should lead to small differences with the master branch, right?

**Review by Thomas-Ulrich — CHANGES_REQUESTED** [review]
I added a few suggestions but overall, this is a nice and readable code!

File: postprocessing/validation/compare-faults/compare.py [hunk 9]
```diff
@@ -49,9 +52,15 @@ def l2_difference(q_0, q_1):
     quantity = fault.ReadData(q, last_index-1)
     quantity_ref = fault_ref.ReadData(q, last_index-1)
     # compute error
-    relative_error = l2_difference(quantity, quantity_ref) / l2_norm(quantity_ref)
-    print(f"{q:3}: {relative_error}")
-    errors[i] = relative_error
+    ref_norm = l2_norm(quantity_ref)
+    if ref_norm < 1e-10:
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply postprocessing/validation/compare-faults/compare.py:56]: I need to deal with quantities, which are supposed to be 0 everywhere. If I check `norm(solution - reference) / norm(reference) != NaN` then, solution can be anything but 0. With this (arbitrary) choice of `1e-10` I try to circumvent this problem. Anyways, I am open to ideas, how to compare to a zero reference.

File: src/DynamicRupture/Typedefs.hpp [hunk 10]
```diff
@@ -30,7 +31,23 @@ struct ImpedancesAndEta {
...
+  real normalStress[CONVERGENCE_ORDER][misc::numPaddedPoints] = {{}};
+  real xyStress[CONVERGENCE_ORDER][misc::numPaddedPoints] = {{}};
+  real xzStress[CONVERGENCE_ORDER][misc::numPaddedPoints] = {{}};
+};
+
+/**
+ * Struct that contains all traction results
+ */
+struct TractionResults {
+  real xyTraction[CONVERGENCE_ORDER][misc::numPaddedPoints] = {{}};
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply src/DynamicRupture/Typedefs.hpp:47]: I agree with updatedTraction 1 and 2. lockedFaultTraction is kind of correct, but in fact it is tractionAsIfFaultWasLocked, isn't it.

File: src/DynamicRupture/FrictionLaws/FastVelocityWeakeningLaw.h [hunk 11]
```diff
@@ -5,132 +5,143 @@
...
+
+    // low-velocity steady state friction coefficient
+    real lowVelocityFriction =
+        this->drParameters.rsF0 -
+        (this->drParameters.rsB - localA) * log(localSlipRate / this->drParameters.rsSr0);
+    real steadyStateFrictionCoefficient =
+        muW + (lowVelocityFriction - muW) /
+                  std::pow(1.0 + misc::power<8>(localSlipRate / localSrW), 1.0 / 8.0);
+    // For compiling reasons we write SINH(X)=(EXP(X)-EXP(-X))/2
+    real steadyStateStateVariable = localA * log(2.0 * this->drParameters.rsSr0 / localSlipRate *
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/FastVelocityWeakeningLaw.h:58]: Yes, you're right. But this is not the asinh, but the sinh here...

File: src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.cpp [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+    real expSigma = std::exp(-hydraulicDiffusivity[ltsFace][pointIndex] * deltaT * tmp);
+
+    // Temperature and pressure diffusion in spectral domain over timestep
+    // This is + F(t) exp(-A dt) in equation (10)
+    real thetaDiffusion = thetaTmpBuffer[ltsFace][pointIndex][tpGridPointIndex] * expTheta;
+    real sigmaDiffusion = sigmaTmpBuffer[ltsFace][pointIndex][tpGridPointIndex] * expSigma;
+
+    // Heat generation during timestep
+    // This is B/A * (1 - exp(-A dt)) in equation (10)
+    // heatSource stores \exp(-\hat{l}^2 / 2)
```
[the remark this answers is not in the document]
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.cpp:93]: True

File: src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h (hunk 8 above)
Comment (sebwolf-de) [reply src/DynamicRupture/FrictionLaws/ThermalPressurization/ThermalPressurization.h:44]: Indeed. In particular the error is in the smallest component:
```
(gdb) inspect disc%dynrup%tp_dfinv
$1 = (1.8863643150858097e-07, 6.6425882217799723e-08, 8.9665562162703227e-08, 1.2103584882157821e-07, 1.6338130656424509e-07, 2.2054169565902143e-07, 2.9770015032307667e-07, 4.0185316993030467e-07, 5.4244504078275839e-07, 7.3222421592652987e-07, 9.8839930698886031e-07, 1.3341995099409315e-06, 1.8009809595573516e-06, 2.4310700105351742e-06, 3.2816012655547887e-06, 4.4296983712616928e-06, 5.9794673613526645e-06, 8.0714366823351432e-06, 1.0895299895442135e-05, 1.4707116525044456e-05, 1.9852530775377855e-05, 2.6798113519817424e-05, 3.6173669561147141e-05, 4.8829346459458476e-05, 6.5912723386480073e-05, 8.8972870194560996e-05, 0.00012010081246744448, 0.00016211913950621388, 0.00021883794833910572, 0.00029540033199740393, 0.00039874873990757139, 0.00053825449857406761, 0.00072656757561761077, 0.00098076364124657532, 0.0013238924392870008, 0.0017870678694548382, 0.0024122893033196959, 0.0032562499625074714, 0.0043954776915597401, 0.0059332742754557525, 0.0080090825384880725, 0.010811130605181308, 0.014593499867258468, 0.019699164329178075, 0.026591090471628043, 0.03589421767617744, 0.048452125871238393, 0.065403528853072282, 0.088285529468870846, 0.11917299953506608, 0.16086672304765989, 0.21714736295177328, 0.29311828042235422, 0.39566829248964758, 0.53409632949570884, 0.72095463446379771, 0.97318696319369502, 1.3136649936851528, 1.7732622619301779, 1.1968268412042979)
(gdb) inspect seissol::dr::friction_law::tpInverseFourierCoefficients
$2 = {values = {2.1324118344448284e-07, 6.6425882217799723e-08, 8.9665562162703227e-08, 1.2103584882157821e-07, 1.6338130656424509e-07, 2.2054169565902143e-07, 2.9770015032307667e-07, 4.0185316993030467e-07, 5.4244504078275839e-07, 7.3222421592652987e-07, 9.8839930698886031e-07, 1.3341995099409315e-06, 1.8009809595573516e-06, 2.4310700105351742e-06, 3.2816012655547887e-06, 4.4296983712616928e-06, 5.9794673613526645e-06, 8.0714366823351432e-06, 1.0895299895442135e-05, 1.4707116525044456e-05, 1.9852530775377855e-05, 2.6798113519817424e-05, 3.6173669561147141e-05, 4.8829346459458476e-05, 6.5912723386480073e-05, 8.8972870194560996e-05, 0.00012010081246744448, 0.00016211913950621388, 0.00021883794833910572, 0.00029540033199740393, 0.00039874873990757139, 0.00053825449857406761, 0.00072656757561761077, 0.00098076364124657532, 0.0013238924392870008, 0.0017870678694548382, 0.0024122893033196959, 0.0032562499625074714, 0.0043954776915597401, 0.0059332742754557525, 0.0080090825384880725, 0.010811130605181308, 0.014593499867258468, 0.019699164329178075, 0.026591090471628043, 0.03589421767617744, 0.048452125871238393, 0.065403528853072282, 0.088285529468870846, 0.11917299953506608, 0.16086672304765989, 0.21714736295177328, 0.29311828042235422, 0.39566829248964758, 0.53409632949570884, 0.72095463446379771, 0.97318696319369502, 1.3136649936851528, 1.7732622619301779, 1.1968268412042979}}
```

**Comment by sebwolf-de** [comment]
After fixing single precision, this PR is ready for review now .@ravil-mobile  and @Thomas-Ulrich already commented. I think I have included most of your changes.

**Review by Thomas-Ulrich — APPROVED** [review]
Looks good.
Please consider updating tractionXY and tractionXZ to traction 1 and 2, e.g.

**Pull request merged**